pages/en/lb4/Testing-Your-Extensions.md

```diff
@@ -0,0 +1,303 @@
+decorator and retrieve the correct metadata. *The retrieval gets tested as a
+consequence of validating the metadata was stored.*
+
+**`src/decorators/test.decorator.ts`**
+```ts
+export function test(file: string) {
+  return function(target: Object, methodName: string): void {
+    Reflector.defineMetadata(
+      'example.msg.decorator.metadata.key',
+      {file},
+      target,
+      methodName,
+    );
+  };
+}
+
+export function getTestMetadata(
+  controllerClass: Constructor<{}>,
+  methodName: string,
+): {file: string} {
+  return Reflector.getMetadata(
+    'example.msg.decorator.metadata.key',
+    controllerClass.prototype,
+    methodName,
+  );
+}
+```
+
+**`test/unit/decorators/test.decorator.ts`**
+```ts
+import {test, getTestMetadata} from '../../..';
+import {expect} from '@loopback/testlab';
+
+describe('test.decorator (unit)', () => {
+  it('can store test name via a decorator', () => {
+    class TestClass {
+      @test('me.test.ts')
+      me() {}
+    }
+
+    const metadata = getTestMetadata(TestClass, 'me');
+    expect(metadata).to.be.a.Object();
+    expect(metadata.file).to.be.eql('me.test.ts');
+  });
+});
+```
+
+#### Mixins
+
+A Mixin is a TypeScript function that extends the `Application` Class, adding
+new constructor properties, methods, etc. It is difficult to write a unit test
+for a Mixin without the `Application` Class dependency. The recommended practice
+is to write an integration test ofr a Mixin as show below.
+
+#### Providers
+
+A Provider is a Class that implements the Provider interface. This interface
+requires the Class to have a `value()` function. A unit test for a provider
+should test the `value()` function by instantiating a new Provider class, using
+a test double for any constructor arguments.
+
+**`src/providers/random-number.provider.ts`**
+```ts
+import {Provider} from '@loopback/context';
+
+export class RandomNumberProvider implements Provider<number> {
+  value(): number {
+    return (max: number): number => {
+      return Math.floor(Math.random() * max) + 1;
+    };
+  }
+}
+```
+
+**`test/unit/providers/random-number.unit.test.ts`**
+```ts
+import {RandomNumberProvider} from '../../..';
+import {expect} from '@loopback/testlab';
+
+describe('RandomNumberProvider (unit)', () => {
+  it('generates a random number within range', () => {
+    const provider = new RandomNumberProvider().value();
+    const random: number = provider(3);
+
+    expect(random).to.be.a.Number();
+    expect(random).to.equalOneOf([1, 2, 3]);
+  })
+})
+```
+
+#### Repositories
+
+*TBD ... How / What?*
+
+### Integration Tests
+
+An integration test plays an important part in your test suite by ensuring your
+extension artifacts play well with each other as well as `@loopback`. It is
+recommended to test two items together and other integrations can be
+substituted by test doubles so it becomes apparent where the integration breaks
+(if it does).
+
+#### Controllers
```

Comment:
maybe we can direct users to http://loopback.io/doc/en/lb4/Testing-your-application.html#integration-testing if we haven't decided the content here.

Reply:
I'm hesitant to do that because integration tests for an extension might be different than those for a application. I also don't want to be too prescriptive because a user should be the one deciding how / what tests they want to write for their extension